tests/testflows/map_type/requirements/requirements.md

```diff
@@ -0,0 +1,523 @@
+
+### Array of Maps
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.ArrayOfMaps
+version: 1.0
+
+[ClickHouse] SHALL support `Array(Map(key, value))` data type.
+
+### Nested With Maps
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.NestedWithMaps
+version: 1.0
+
+[ClickHouse] SHALL support defining `Map(key, value)` data type inside the [Nested] data type.
+
+### Value Retrieval
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Value.Retrieval
+version: 1.0
+
+[ClickHouse] SHALL support getting the value from a `Map(key, value)` data type using `map[key]` syntax.
+If `key` has duplicates then the first `key:value` pair SHALL be returned. 
+
+For example,
+
+```sql
+SELECT a['key2'] FROM table_map;
+```
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Value.Retrieval.KeyInvalid
+version: 1.0
+
+[ClickHouse] SHALL return an error when key does not match the key type.
+
+For example,
+
+```sql
+SELECT map(1,2) AS m, m[1024]
+```
+
+SHALL have the following output 
+
+```bash
+Code: 43. DB::Exception: Received from localhost:9000. DB::Exception: Illegal types of arguments: Map(UInt8,UInt8), UInt16 for function : While processing {1, 2} AS m, m[1024]. 
+```
+
+Exceptions:
+
+* when key is `NULL` the return value SHALL be `NULL`
+* when key has [Integer] type and key is either too large or negative then
+  when reading from a table column it SHALL return `zero`.
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Value.Retrieval.KeyNotFound
+version: 1.0
+
+[ClickHouse] SHALL return
+
+* zeros for [Integer] values
+* empty strings for [String] values
+* empty for [Array] values
+
+when there's no corresponding `key` defined in the `Map(key, value)` data type. 
+
+
+### Converting Tuple(Array, Array) to Map
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Conversion.From.TupleOfArraysToMap
+version: 1.0
+
+[ClickHouse] SHALL support converting [Tuple(Array, Array)] to `Map(key, value)` using the [CAST] function.
+
+``` sql
+SELECT CAST(([1, 2, 3], ['Ready', 'Steady', 'Go']), 'Map(UInt8, String)') AS map;
+```
+
+``` text
+┌─map───────────────────────────┐
+│ {1:'Ready',2:'Steady',3:'Go'} │
+└───────────────────────────────┘
+```
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Conversion.From.TupleOfArraysMap.Invalid
+version: 1.0
+
+[ClickHouse] SHALL return an error when casting [Tuple(Array, Array)] when arguments are invalid
+such as when arrays are not of equal size.
+
+For example,
+
+```sql
+SELECT CAST(([2, 1, 1023], ['', '']), 'Map(UInt8, String)') AS map, map[10]
+```
+
+### Converting Array(Tuple(K,V)) to Map
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Conversion.From.ArrayOfTuplesToMap
+version: 1.0
+
+[ClickHouse] SHALL support converting [Array(Tuple(K,V))] to `Map(key, value)` using the [CAST] function.
+
+For example,
+
+```sql
+SELECT CAST(([(1,2),(3)]), 'Map(UInt8, UInt8)') AS map
+```
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Conversion.From.ArrayOfTuplesToMap.Invalid
+version: 1.0
+
+[ClickHouse] SHALL return an error when casting [Array(Tuple(K, V))] to `Map(key, value)`
```

Comment:
I don't like this behaviour.

Reply:
Ok, but this is how it behaves right now. When it changes we can update the requirement.

Reply:
Updated requirement to use MAY.